internal/provider/resource_apply_module.go

```diff
@@ -0,0 +1,589 @@
+	}
+
+	plannedRun, err := t.client.Run.GetRun(ctx, &sdktypes.GetRunInput{ID: createdRun.Metadata.ID})
+	if err != nil {
+		diags.AddError("Failed to get planned run", err.Error())
+		return diags
+	}
+
+	// If the plan fails, both plannedRun.Status and plannedRun.Plan.Status are "errored".
+	// If the plan succeeds, plannedRun.Status is "planned",
+	// while plannedRun.Plan.Status is "finished".
+	//
+	if !strings.HasPrefix(string(plannedRun.Status), "planned") {
+		diags.AddError("Plan failed", string(plannedRun.Status))
+		return diags
+	}
+	if plannedRun.Plan.Status != "finished" {
+		diags.AddError("Plan failed", string(plannedRun.Plan.Status))
+		return diags
+	}
+
+	// Capture the run ID.
+	runID := plannedRun.Metadata.ID
+
+	// TODO: Also take this early return when the API and SDK support speculative runs and PlanOnly is implemented.
+
+	if plannedRun.Status == "planned_and_finished" {
+		// Return the output.
+		// Leave the ID field unknown here.
+		output.WorkspacePath = types.StringValue(plannedRun.WorkspacePath)
+		output.ModuleSource = types.StringValue(*plannedRun.ModuleSource)
+		output.ModuleVersion = types.StringValue(*plannedRun.ModuleVersion)
+		output.Variables = input.model.Variables // Cannot get variables back from a workspace or run, so pass them through.
+		return nil
+	}
+
+	// Do the apply run.
+	appliedRun, err := t.client.Run.ApplyRun(ctx, &sdktypes.ApplyRunInput{
+		RunID:   runID,
+		Comment: &applyRunComment,
+	})
+	if err != nil {
+		diags.AddError("Failed to apply a run", err.Error())
+		return diags
+	}
+
+	// Make sure the run has an apply.
+	if appliedRun.Apply == nil {
+		msg := fmt.Sprintf("Created run does not have an apply: %s", appliedRun.Metadata.ID)
+		diags.AddError(msg, "")
+		return diags
+	}
+
+	if err = t.waitForJobCompletion(ctx, appliedRun.Apply.CurrentJobID); err != nil {
+		diags.AddError("Failed to wait for apply job completion", err.Error())
+		return diags
+	}
+
+	finishedRun, err := t.client.Run.GetRun(ctx, &sdktypes.GetRunInput{ID: appliedRun.Metadata.ID})
+	if err != nil {
+		diags.AddError("Failed to get finished run", err.Error())
+		return diags
+	}
+
+	// If an apply job succeeds, finishedRun.Status is "applied" and
+	// finishedRun.Apply.Status is "finished".
+	if finishedRun.Status != "applied" {
+		diags.AddError("Apply failed", string(finishedRun.Status))
+		return diags
+	}
+	if finishedRun.Apply.Status != "finished" {
+		diags.AddError("Apply status", string(finishedRun.Apply.Status))
+		return diags
+	}
+
+	// In case of a rainy day, make sure the ModuleSource and ModuleVersion *string aren't nil.
+	if finishedRun.ModuleSource == nil {
+		diags.AddError("Finished run's module source is nil.", "")
+		return diags
+	}
+	if finishedRun.ModuleVersion == nil {
+		diags.AddError("Finished run's module version is nil.", "")
+		return diags
+	}
+
+	// Return the output.
+	// Leave the ID field unknown here.
+	output.WorkspacePath = types.StringValue(finishedRun.WorkspacePath)
```

Comment:
It doesn't seem like this output should be needed since it'll always be the same as the input if no errors occurred during the run.